tao/tao_bridge/passes/tao_mark_for_compilation_pass.cc

```diff
@@ -1507,6 +1510,65 @@ Status MarkForCompilationPassImpl::FindCompilationCandidates() {
       }
     }
 
+    // This is used to fix some corner cases of our clustering strategy.
+    // We roughly divide the operands of an op into two categories: data operand
+    // and shape operand. The shape operand (e.g. the first operand of tf.Fill
+    // op) does not affect the content of the outputs of the op and is only used
+    // to calcaulate the rank and shape of the outputs. Because BladeDISC
+    // requires all tensors have static rank, the shape operand (and the
+    // subgraph used to calcaulate it) should be fix shaped or a const,
+    // depending on the rank of the output is affected by the shape or the
+    // content of the shape operand.
+    //
+    // The subgraph used to compute the shape operands is usually cheap and
+    // simple. Thus the above requirement is not a big problem for the most of
+    // common cases. However, we do find that seme shape-operand-related
+    // subgraphs can be very complicated and expensive in some cases.
+    //   An Example from a real TTS model:
+    //    ```
+    //    ...
+    //    %tx = tf.Matmul(...)
+    //    ...
+    //    %ty = tf.Max(use(%tx))
+    //    %shape = tf.pack(%ty, ...)
+    //    %out = tf.Fill(%shape, ...)
+    //    ```
+    // If we group the `tf.Fill` and the related subgraph into the same cluster,
+    // the whole cluster will have to be fix-shaped, leading to unnecessary
+    // re-compilation.
+    //
+    // In this workaround, we try to find such `tf.Fill` ops and exclude them
+    // from compilation to make sure the remaining shape-operand-related-graph
+    // only needs to be compiled once.
+    //
+    // TODO(disc): support other ops having shape operands.
+    if (debug_options_.cluster_for_mlir && node->type_string() == "Fill" &&
+        !compile_time_fixed_shape_nodes[node->id()]) {
+      std::unordered_set<Node*> visitedNodeSet;
+      auto hasExpensiveProducers = [&](Node* node) {
+        std::vector<Node*> node_queue{node};
+        while (!node_queue.empty()) {
+          Node* cur = node_queue.back();
+          node_queue.pop_back();
+          if (!visitedNodeSet.insert(cur).second) continue;
+          for (auto e : cur->in_edges()) {
+            if (e->IsControlEdge()) continue;
+            Node* src = e->src();
+            if (!compile_time_fixed_shape_nodes[src->id()]) continue;
+            if (src->type_string() == "Max") return true;
```

Comment:
I weakly prefer to use 'MatMul' to define 'expensive' rather than 'Max' which is more explainable.

Reply:
done